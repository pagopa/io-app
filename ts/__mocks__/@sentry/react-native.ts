export const captureException = jest.fn();
export const captureMessage = jest.fn();
export const captureEvent = jest.fn();
export const setContext = jest.fn();
export const setExtras = jest.fn();
export const setExtra = jest.fn();
export const setTags = jest.fn();
export const setTag = jest.fn();
export const setUser = jest.fn();
export const lastEventId = jest.fn();
export const captureCheckIn = jest.fn();
export const withMonitor = jest.fn();
export const flush = jest.fn().mockResolvedValue(false);
export const close = jest.fn().mockResolvedValue(false);
export const isInitialized = jest.fn().mockReturnValue(false);
export const isEnabled = jest.fn().mockReturnValue(false);
export const addEventProcessor = jest.fn();
export const startSession = jest.fn();
export const endSession = jest.fn();
export const captureSession = jest.fn();
export const ErrorBoundary = jest.fn(({ children }) => children);
