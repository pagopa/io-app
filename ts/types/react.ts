import { Component, ComponentType, StatelessComponent } from "react";
import {
  NavigationInjectedProps,
  NavigationScreenProps
} from "react-navigation";

/**
 * Evaluates to the Props type of a React component
 */
export type ComponentProps<C> = C extends StatelessComponent<infer P1>
  ? P1
  : C extends Component<infer P2>
    ? P2
    : C extends ComponentType<infer P3> ? P3 : never;

/**
 * Infers the type of the navigation params of a component
 */
export type InferNavigationParams<
  C,
  P = ComponentProps<C>
> = P extends NavigationInjectedProps<infer N>
  ? N
  : P extends NavigationScreenProps<infer N1> ? N1 : never;
