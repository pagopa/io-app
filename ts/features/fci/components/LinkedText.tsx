import * as React from "react";
import * as O from "fp-ts/lib/Option";
import { H4 } from "../../../components/core/typography/H4";
import { Link } from "../../../components/core/typography/Link";
import { WithTestID } from "../../../types/WithTestID";

type Props = WithTestID<{
  text: string;
  replacementUrl: string;
  onPress: (holder: string) => void;
}>;

// a regex to match markdown links
const matchMarkdownLink = /!?\[([^\]]*)\]\(([^\\)]+)\)/g;

// a function to get all the matched links
const getMatchedLinks = (text: string) => text.match(matchMarkdownLink) ?? [];

/**
 * This function returns the replacement url
 * for the tag DOCUMENT_URL and PRIVACY_URL
 * otherwise returns the tag because
 * it's a custom link
 */
const getOrReplaceTagWithLink = (tagLink: string, replacementUrl: string) => {
  switch (tagLink) {
    case "@DOCUMENT_URL":
    case "@PRIVACY_URL":
      return replacementUrl;
    default:
      return tagLink;
  }
};

/**
 * This component renders a text with markdown links
 */
const LinkedText = (props: Props) => {
  const parts = getMatchedLinks(props.text);
  // generate unique string with the separator
  const partsToString = parts.join("|");
  // escape special characters
  const stringReg = partsToString.replace(/[[\\\]()]/g, "\\$&");
  // replace the markdown link with a placeholder
  // and use the separator to split the text
  const replaceLinksWithPlaceholder = new RegExp(stringReg, "g");
  const textWithSeparator = props.text.replace(
    replaceLinksWithPlaceholder,
    "$@"
  );

  /**
   * This function generates an array of Link components
   * used to render the list of clauses
   */
  const generateArrayOfLinkedText = (
    text: string,
    onPress: (holder: string) => void
  ) => {
    const parts = getMatchedLinks(text);

    // a function to render the link
    const getTextWithLinkComponent = (matched: string, index: number) => {
      const splitted = matched.split(matchMarkdownLink);
      const textToBeLinked = splitted[1];
      const url = splitted[2];
      return (
        <Link
          key={index}
          onPress={() =>
            onPress(getOrReplaceTagWithLink(url, props.replacementUrl))
          }
        >
          {textToBeLinked}
        </Link>
      );
    };

    return parts.map(
      (part, index) => part && getTextWithLinkComponent(part, index)
    );
  };

  const arrayOfLinkedText = generateArrayOfLinkedText(
    props.text,
    props.onPress
  );

  return (
    <H4>
      {textWithSeparator.split("$@").map((text, index) =>
        O.isSome(O.fromNullable(arrayOfLinkedText[index])) ? (
          <>
            <H4 weight={"Regular"} color={"bluegreyDark"}>
              {text}
            </H4>
            {arrayOfLinkedText[index]}
          </>
        ) : (
          <H4 weight={"Regular"} color={"bluegreyDark"}>
            {text}
          </H4>
        )
      )}
    </H4>
  );
};

export default LinkedText;
