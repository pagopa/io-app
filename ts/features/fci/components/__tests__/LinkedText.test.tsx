import * as React from "react";
import { render } from "@testing-library/react-native";
import LinkedText from "../LinkedText";

type Props = {
  text: string;
  replacementUrl: string;
  onPress: (holder: string) => void;
};

const MOCK_TEXT_ONE_LINK =
  "This is a [linked text](@DOCUMENT_URL) to be clicked";
const MOCK_TEXT_MULTIPLE_LINK =
  "This is a [linked text1](@DOCUMENT_URL) to be clicked. This is a [linked text2](@DOCUMENT_URL) to be clicked.";
const MOCK_TEXT_CUSTOM_LINK =
  "This is a [linked text1](@DOCUMENT_URL) to be clicked. This is a [linked text2](https://fakeUrl.com) to be clicked.";

describe("Test LinkedText component", () => {
  it("should render a LinkedText component with props correctly", () => {
    const props = {
      text: "Clause title 1",
      replacementUrl: "https://fakeUrl.com",
      onPress: jest.fn()
    };
    const component = renderComponent({ ...props });
    expect(component).toBeTruthy();
    expect(component).toMatchSnapshot();
  });
  it("should render a LinkedText component correctly with one link", () => {
    const props = {
      text: MOCK_TEXT_ONE_LINK,
      replacementUrl: "https://fakeUrl.com",
      onPress: jest.fn()
    };
    const component = renderComponent({ ...props });
    expect(component).toBeTruthy();
    expect(component.queryByText("linked text")).toBeTruthy();
  });
  it("should render a LinkedText component correctly with text and multiple link", () => {
    const props = {
      text: MOCK_TEXT_MULTIPLE_LINK,
      replacementUrl: "https://fakeUrl.com",
      onPress: jest.fn()
    };
    const component = renderComponent({ ...props });
    expect(component).toBeTruthy();
    expect(component.queryByText("linked text1")).toBeTruthy();
    expect(component.queryByText("linked text2")).toBeTruthy();
  });
  it("should render a LinkedText component correctly with text and multiple custom link", () => {
    const onPressDetail = jest.fn();
    const props = {
      text: MOCK_TEXT_CUSTOM_LINK,
      replacementUrl: "https://fakeUrl.com",
      onPress: onPressDetail
    };
    const component = renderComponent({ ...props });
    expect(component).toBeTruthy();
    expect(component.queryByText("linked text1")).toBeTruthy();
    expect(component.queryByText("linked text2")).toBeTruthy();
  });
});

const renderComponent = (props: Props) => render(<LinkedText {...props} />);
