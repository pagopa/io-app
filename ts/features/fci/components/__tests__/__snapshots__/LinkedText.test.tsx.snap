// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test LinkedText component should render a LinkedText component with props correctly 1`] = `
<Text
  color="bluegreyDark"
  font="TitilliumWeb"
  fontStyle={
    Object {
      "fontSize": 16,
      "lineHeight": 22,
    }
  }
  style={
    Array [
      Object {
        "fontSize": 16,
        "lineHeight": 22,
      },
      Object {
        "color": "#17324D",
        "fontFamily": "Titillium Web",
        "fontStyle": "normal",
        "fontWeight": "700",
      },
    ]
  }
  weight="Bold"
  weightColorFactory={[Function]}
>
  <Text
    color="bluegreyDark"
    font="TitilliumWeb"
    fontStyle={
      Object {
        "fontSize": 16,
        "lineHeight": 22,
      }
    }
    style={
      Array [
        Object {
          "fontSize": 16,
          "lineHeight": 22,
        },
        Object {
          "color": "#17324D",
          "fontFamily": "Titillium Web",
          "fontStyle": "normal",
          "fontWeight": "400",
        },
      ]
    }
    weight="Regular"
    weightColorFactory={[Function]}
  />
  <Text
    color="bluegreyDark"
    font="TitilliumWeb"
    fontStyle={
      Object {
        "fontSize": 16,
        "lineHeight": 22,
      }
    }
    style={
      Array [
        Object {
          "fontSize": 16,
          "lineHeight": 22,
        },
        Object {
          "color": "#17324D",
          "fontFamily": "Titillium Web",
          "fontStyle": "normal",
          "fontWeight": "400",
        },
      ]
    }
    weight="Regular"
    weightColorFactory={[Function]}
  >
    C
  </Text>
  <Text
    color="bluegreyDark"
    font="TitilliumWeb"
    fontStyle={
      Object {
        "fontSize": 16,
        "lineHeight": 22,
      }
    }
    style={
      Array [
        Object {
          "fontSize": 16,
          "lineHeight": 22,
        },
        Object {
          "color": "#17324D",
          "fontFamily": "Titillium Web",
          "fontStyle": "normal",
          "fontWeight": "400",
        },
      ]
    }
    weight="Regular"
    weightColorFactory={[Function]}
  >
    l
  </Text>
  <Text
    color="bluegreyDark"
    font="TitilliumWeb"
    fontStyle={
      Object {
        "fontSize": 16,
        "lineHeight": 22,
      }
    }
    style={
      Array [
        Object {
          "fontSize": 16,
          "lineHeight": 22,
        },
        Object {
          "color": "#17324D",
          "fontFamily": "Titillium Web",
          "fontStyle": "normal",
          "fontWeight": "400",
        },
      ]
    }
    weight="Regular"
    weightColorFactory={[Function]}
  >
    a
  </Text>
  <Text
    color="bluegreyDark"
    font="TitilliumWeb"
    fontStyle={
      Object {
        "fontSize": 16,
        "lineHeight": 22,
      }
    }
    style={
      Array [
        Object {
          "fontSize": 16,
          "lineHeight": 22,
        },
        Object {
          "color": "#17324D",
          "fontFamily": "Titillium Web",
          "fontStyle": "normal",
          "fontWeight": "400",
        },
      ]
    }
    weight="Regular"
    weightColorFactory={[Function]}
  >
    u
  </Text>
  <Text
    color="bluegreyDark"
    font="TitilliumWeb"
    fontStyle={
      Object {
        "fontSize": 16,
        "lineHeight": 22,
      }
    }
    style={
      Array [
        Object {
          "fontSize": 16,
          "lineHeight": 22,
        },
        Object {
          "color": "#17324D",
          "fontFamily": "Titillium Web",
          "fontStyle": "normal",
          "fontWeight": "400",
        },
      ]
    }
    weight="Regular"
    weightColorFactory={[Function]}
  >
    s
  </Text>
  <Text
    color="bluegreyDark"
    font="TitilliumWeb"
    fontStyle={
      Object {
        "fontSize": 16,
        "lineHeight": 22,
      }
    }
    style={
      Array [
        Object {
          "fontSize": 16,
          "lineHeight": 22,
        },
        Object {
          "color": "#17324D",
          "fontFamily": "Titillium Web",
          "fontStyle": "normal",
          "fontWeight": "400",
        },
      ]
    }
    weight="Regular"
    weightColorFactory={[Function]}
  >
    e
  </Text>
  <Text
    color="bluegreyDark"
    font="TitilliumWeb"
    fontStyle={
      Object {
        "fontSize": 16,
        "lineHeight": 22,
      }
    }
    style={
      Array [
        Object {
          "fontSize": 16,
          "lineHeight": 22,
        },
        Object {
          "color": "#17324D",
          "fontFamily": "Titillium Web",
          "fontStyle": "normal",
          "fontWeight": "400",
        },
      ]
    }
    weight="Regular"
    weightColorFactory={[Function]}
  >
     
  </Text>
  <Text
    color="bluegreyDark"
    font="TitilliumWeb"
    fontStyle={
      Object {
        "fontSize": 16,
        "lineHeight": 22,
      }
    }
    style={
      Array [
        Object {
          "fontSize": 16,
          "lineHeight": 22,
        },
        Object {
          "color": "#17324D",
          "fontFamily": "Titillium Web",
          "fontStyle": "normal",
          "fontWeight": "400",
        },
      ]
    }
    weight="Regular"
    weightColorFactory={[Function]}
  >
    t
  </Text>
  <Text
    color="bluegreyDark"
    font="TitilliumWeb"
    fontStyle={
      Object {
        "fontSize": 16,
        "lineHeight": 22,
      }
    }
    style={
      Array [
        Object {
          "fontSize": 16,
          "lineHeight": 22,
        },
        Object {
          "color": "#17324D",
          "fontFamily": "Titillium Web",
          "fontStyle": "normal",
          "fontWeight": "400",
        },
      ]
    }
    weight="Regular"
    weightColorFactory={[Function]}
  >
    i
  </Text>
  <Text
    color="bluegreyDark"
    font="TitilliumWeb"
    fontStyle={
      Object {
        "fontSize": 16,
        "lineHeight": 22,
      }
    }
    style={
      Array [
        Object {
          "fontSize": 16,
          "lineHeight": 22,
        },
        Object {
          "color": "#17324D",
          "fontFamily": "Titillium Web",
          "fontStyle": "normal",
          "fontWeight": "400",
        },
      ]
    }
    weight="Regular"
    weightColorFactory={[Function]}
  >
    t
  </Text>
  <Text
    color="bluegreyDark"
    font="TitilliumWeb"
    fontStyle={
      Object {
        "fontSize": 16,
        "lineHeight": 22,
      }
    }
    style={
      Array [
        Object {
          "fontSize": 16,
          "lineHeight": 22,
        },
        Object {
          "color": "#17324D",
          "fontFamily": "Titillium Web",
          "fontStyle": "normal",
          "fontWeight": "400",
        },
      ]
    }
    weight="Regular"
    weightColorFactory={[Function]}
  >
    l
  </Text>
  <Text
    color="bluegreyDark"
    font="TitilliumWeb"
    fontStyle={
      Object {
        "fontSize": 16,
        "lineHeight": 22,
      }
    }
    style={
      Array [
        Object {
          "fontSize": 16,
          "lineHeight": 22,
        },
        Object {
          "color": "#17324D",
          "fontFamily": "Titillium Web",
          "fontStyle": "normal",
          "fontWeight": "400",
        },
      ]
    }
    weight="Regular"
    weightColorFactory={[Function]}
  >
    e
  </Text>
  <Text
    color="bluegreyDark"
    font="TitilliumWeb"
    fontStyle={
      Object {
        "fontSize": 16,
        "lineHeight": 22,
      }
    }
    style={
      Array [
        Object {
          "fontSize": 16,
          "lineHeight": 22,
        },
        Object {
          "color": "#17324D",
          "fontFamily": "Titillium Web",
          "fontStyle": "normal",
          "fontWeight": "400",
        },
      ]
    }
    weight="Regular"
    weightColorFactory={[Function]}
  >
     
  </Text>
  <Text
    color="bluegreyDark"
    font="TitilliumWeb"
    fontStyle={
      Object {
        "fontSize": 16,
        "lineHeight": 22,
      }
    }
    style={
      Array [
        Object {
          "fontSize": 16,
          "lineHeight": 22,
        },
        Object {
          "color": "#17324D",
          "fontFamily": "Titillium Web",
          "fontStyle": "normal",
          "fontWeight": "400",
        },
      ]
    }
    weight="Regular"
    weightColorFactory={[Function]}
  >
    1
  </Text>
  <Text
    color="bluegreyDark"
    font="TitilliumWeb"
    fontStyle={
      Object {
        "fontSize": 16,
        "lineHeight": 22,
      }
    }
    style={
      Array [
        Object {
          "fontSize": 16,
          "lineHeight": 22,
        },
        Object {
          "color": "#17324D",
          "fontFamily": "Titillium Web",
          "fontStyle": "normal",
          "fontWeight": "400",
        },
      ]
    }
    weight="Regular"
    weightColorFactory={[Function]}
  />
</Text>
`;
